File: consoles/hac.py
from pathlib import Path

from console import Console
from error import Error


class Hac(Console):
    BinTool = 'elf2nro "{elf}" "{output}.nro" --nacp="{nacp}" --icon="{icon}" --romfs="{romfs}"'
    NacpTool = 'nacptool --create "{name}" "{author}" "{version}" "{out}.nacp"'

    def __init__(self, metadata: dict) -> None:
        super().__init__(metadata)

    def build(self, build_dir: Path) -> str | Error:
        args = {
            "name": self.title,
            "author": self.author,
            "version": self.version,
            "out": build_dir / self.title,
        }

        if (value := self.run_command(Hac.NacpTool, args)) != Error.NONE:
            return value

        args = {
            "nacp": build_dir / f"{self.title}.nacp",
            "icon": self.icon_file(),
            "romfs": self.path_to("files.romfs"),
            "elf": self.binary_path(),
            "output": build_dir / self.title,
        }

        if (value := self.run_command(Hac.BinTool, args)) != Error.NONE:
            return value

        with open(self.final_binary_path(build_dir), "ab") as executable:
            executable.write(self.game_zip.read_bytes())

        return Error.NONE

    def binary_extension(self) -> str:
        return "nro"

    def icon_extension(self) -> str:
        return "jpg"

File: backend/src/types/console/hac.rs
use std::path::Path;

use crate::enums::api_error::ApiError;
use crate::enums::console::Console;
use crate::modules::filesystem;
use crate::traits::binary::Binary;
use crate::types::bundler_query::Metadata;

pub struct Hac {}

const METADATA_TOOL: &str = "nacptool";
const BINARY_TOOL: &str = "elf2nro";

impl Binary for Hac {
    async fn compile(
        basepath: &Path,
        metadata: &Metadata,
        icon: &String,
    ) -> Result<String, ApiError> {
        let output_path = filesystem::normalize_path(basepath);

        Self::run_command(
            METADATA_TOOL,
            &[
                "--create",
                &metadata.title,
                &metadata.author,
                &metadata.version,
                &format!("{}.smdh", output_path),
            ],
        )
        .await?;

        let binary_path = Console::Hac.get_binary();

        Self::run_command(
            BINARY_TOOL,
            &[
                &binary_path,
                &format!("{}.nro", output_path),
                &format!("--icon={}", icon),
                &format!("--nacp={}.nacp", output_path),
                &format!("--romfs={}", Console::Hac.get_romfs()),
            ],
        )
        .await?;

        let base64 = filesystem::read_file_as_base64(&format!("{}.nro", output_path)).await?;

        Ok(base64)
    }
}
